### PR TITLE
Feature object modification during the export

### PR DESCRIPTION
**Pull request by przemkalit**
# What does this PR do?
This PR provides a solution for use cases that require modification of objects while they are being exported to a files. In this change, I am introducing the usage of two dictionaries: static and dynamic.

* The static dictionary allows modification of variables that will be changed for every object of a certain kind.
* The dynamic dictionary allows modification of variables that will be changed for a defined object of a certain kind.

remark: the dynamic structure has a precedence over static.

Not all parameters are allowed to be changed due to the possibility of duplication after import.

We've been in touch with @Tompage1994 about this solution.

# How should this be tested?
1) Create structures like below:
```yaml
template_overrides_resources:
  job_template:
    JT1_name:
      description: "JT1_description"
    JT2_name:
      description: "JT2_description"

template_overrides_global:
  job_template:
    scm_branch: "master"
```
2) Use it with filetree_create.

# Is there a relevant Issue open for this?

N/A
# Other Relevant info, PRs, etc
N/A

### CODE REVIEW

**Review by Tompage1994 — CHANGES_REQUESTED** [review]
Generally looks good to me. This shouldn't cause any breaking changes so I'm happy enough with it. I think it does constitute a major change, though. 

One thing I would like to see is a little more documentation on this. @przemkalit could you add in an example in the readme to show how you would use it please?

**Comment by przemkalit** [comment]
Ok. I've changed the name of variables to more meaningful and provided an example in README.

**Review by Tompage1994 — APPROVED** [review]
Perfect, thank you. 

This is even better than I imagined when you described to me so thanks a lot for this contribution!

**Comment by Tompage1994** [comment]
@adonisgarciac @silvinux @ivarmu - You guys are the owners of this role and this is a big change so I want to make sure you're happy with it.

File: roles/filetree_create/templates/current_job_templates.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_job_templates.j2]: This `if` seems to be incomplete.

File: roles/filetree_create/templates/current_job_templates.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_job_templates.j2]: This `if` seems to be incomplete.

File: roles/filetree_create/templates/current_job_templates.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_job_templates.j2]: `extra_var` should be `extra_vars` for naming consistency.

File: roles/filetree_create/templates/current_job_templates.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_job_templates.j2]: `extra_var` should be `extra_vars` for naming consistency.

File: roles/filetree_create/templates/current_projects.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_projects.j2]: Does it make any difference? It looks uglier than the original version and it's not introducing any benefit here.

File: roles/filetree_create/templates/current_projects.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_projects.j2]: Does it make any difference? It looks uglier than the original version and it's not introducing any benefit here.

File: roles/filetree_create/templates/current_projects.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_projects.j2]: Does it make any difference? It looks uglier than the original version and it's not introducing any benefit here.

File: roles/filetree_create/templates/current_schedules.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_schedules.j2:26]: `instance_groups` are a very adequate thing to be customized meanwhile the export process is executed... is there any reason to avoid this setting to be customized?

File: roles/filetree_create/templates/current_workflow_job_templates.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_workflow_job_templates.j2]: Isn't this change introducing extra spaces in the rendered output file?

File: roles/filetree_create/templates/current_inventory_sources.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_inventory_sources.j2]: This `if` seems to be incomplete.

File: roles/filetree_create/templates/current_inventory_sources.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_inventory_sources.j2]: Did you tested that change? It is not related with the PR itself, so this is why i'm asking...

File: roles/filetree_create/templates/current_inventories.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_inventories.j2]: Did you tested that change? It is not related with the PR itself, so this is why i'm asking...

File: roles/filetree_create/templates/current_inventories.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_inventories.j2]: This `if` seems to be incomplete.

File: roles/filetree_create/templates/current_groups.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_groups.j2]: `inventory.nam` should be `inventory.name`

File: roles/filetree_create/templates/current_credentials.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_credentials.j2]: It would make sense to let only from `template_overrides_resources` instead of only from `template_overrides_global`...

File: roles/filetree_create/templates/current_credentials.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_credentials.j2]: This is the only place `organization` is allowed to be overriden... and it comes only from `template_overrides_global`... I'm not sure if it makes sense...

File: roles/filetree_create/templates/current_credentials.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_credentials.j2]: Makes it sense setting all the exported `credentials` with the same (`template_overrides_global`) inputs?

File: roles/filetree_create/templates/current_credentials.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_credentials.j2]: Makes it sense setting all the exported `credentials` with the same (`template_overrides_global`) inputs?

File: roles/filetree_create/templates/current_credential_types.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_credential_types.j2:6]: Makes it sense setting all the exported `credential_types` with the same (`template_overrides_global`) description?

File: roles/filetree_create/templates/current_credential_types.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_credential_types.j2]: Makes it sense setting all the exported `credential_types` with the same (`template_overrides_global`) inputs?

File: roles/filetree_create/templates/current_credential_types.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_credential_types.j2:17]: Makes it sense setting all the exported `credential_types` with the same (`template_overrides_global`) injectors?

File: roles/filetree_create/templates/current_applications.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_applications.j2:7]: Makes it sense setting all the exported `applications` with the same (`template_overrides_global`) description? **This applies to the description of all the other modified objects.**

**Review by ivarmu — CHANGES_REQUESTED** [review]
This is a big change. I'm still thinking about this, as the initial intention for the `filetree_create` role was to export the exact configuration that was applied in a given AWX/Tower/Controller instance... 

@adonisgarciac and @silvinux, what's your opinion about this?

In the meantime, the changes requested should be managed.

File: roles/filetree_create/templates/current_inventory_sources.j2
Comment (przemkalit) [reply roles/filetree_create/templates/current_inventory_sources.j2]: Change done by mistake, reverted.

File: roles/filetree_create/templates/current_inventories.j2
Comment (przemkalit) [reply roles/filetree_create/templates/current_inventories.j2]: Change done by mistake, reverted

File: roles/filetree_create/README.md
Comment (adonisgarciac) [inline roles/filetree_create/README.md]: @Tompage1994  should we use new collection name? aap_configuration_extended

**Review by adonisgarciac — COMMENTED** [review]
Filetree_create was meant to export the exact content that we had in controller/tower/awx. In any case, I think this should not break anything and it will give people the opportunity to override some of the content during export time.

In any case, as @ivarmu has mentioned in several times, I don't like the global option that much. I'm not able to see any use case to have a global value for a parameter in every object. Can you @przemkalit explain the use case for a global value?

File: roles/filetree_create/README.md
Comment (Tompage1994) [reply roles/filetree_create/README.md]: Yes, it should be. Before release of this collection (Which I guess we will do around the same time we get to releasing the aap_configuration we will ensure that the standards of this are met, but this should be changed

Comment (Tompage1994) [inline roles/filetree_create/README.md]:
```suggestion
    - infra. aap_configuration_extended.filetree_create
```

File: roles/filetree_create/templates/current_credential_types.j2
Comment (Tompage1994) [reply roles/filetree_create/templates/current_credential_types.j2:6]: I guess unlikely, but you MIGHT want to (which is what this gives the option to do. For example you might want to just set the description across all cred_types saying that this is added by duplication from a primary cluster

File: roles/filetree_create/templates/current_credential_types.j2
Comment (Tompage1994) [reply roles/filetree_create/templates/current_credential_types.j2]: Almost certainly no for this. But it doesn't harm leaving the option I suppose

**Comment by Tompage1994** [comment]
> Filetree_create was meant to export the exact content that we had in controller/tower/awx. In any case, I think this should not break anything and it will give people the opportunity to override some of the content during export time.
> 
> In any case, as @ivarmu has mentioned in several times, I don't like the global option that much. I'm not able to see any use case to have a global value for a parameter in every object. Can you @przemkalit explain the use case for a global value?

I can see the use case for globlal option for a lot of things.

The key example given to me was that for all projects imported you might want to change to a different branch (e.g. if you're doing this for a promotion process between lifecycle environments. Another example (which is very common) is to disable schedules when importing to a secondary cluster.

I definitely see its value, however I'm not sure it would ever be used on ALL of the options, but my thoughts are that it doesn't harm to allow it to be set in case someone has a weird use case where they want to override certain aspects for some reason

File: roles/filetree_create/README.md
Comment (djdanielsson) [reply roles/filetree_create/README.md]: my PR #15  tried to fix most of these

**Review by adonisgarciac — APPROVED** [review]

**Comment by adonisgarciac** [comment]
Thanks @Tompage1994 for explanation. LGTM then, wait for @ivarmu approval

File: roles/filetree_create/README.md
Comment (ivarmu) [reply roles/filetree_create/README.md]: I'm not sure if this change has been implemented or not (I'm on holiday and writing from telephone). In the proposal, I believe there's an extra space.

**Comment by silvinux** [comment]
While this role wasn't initially designed to support on-the-fly changes, I think it's a valuable feature. It allows you to export configurations, which can be beneficial in other environments or for individuals who frequently perform demos, as it offers flexibility to quickly modify certain values. LGTM!

File: roles/filetree_create/templates/current_applications.j2
Comment (ivarmu) [reply roles/filetree_create/templates/current_applications.j2:7]: This should be resolved.

File: roles/filetree_create/templates/current_credential_types.j2
Comment (ivarmu) [reply roles/filetree_create/templates/current_credential_types.j2:6]: I would agree if the same description should be added to the exported one, but the current proposal is setting the exact same description to every credential type, what have nonsense in my opinion.

File: roles/filetree_create/templates/current_credential_types.j2
Comment (ivarmu) [reply roles/filetree_create/templates/current_credential_types.j2]: It can lead to more errors than necessary... I'd prefer not to let this kind of modifications.

File: roles/filetree_create/templates/current_credential_types.j2
Comment (ivarmu) [reply roles/filetree_create/templates/current_credential_types.j2:17]: Same as before, I'd prefer not to let doing that kind of things.

File: roles/filetree_create/templates/current_credentials.j2
Comment (ivarmu) [reply roles/filetree_create/templates/current_credentials.j2]: This should be resolved.

File: roles/filetree_create/templates/current_credentials.j2
Comment (ivarmu) [reply roles/filetree_create/templates/current_credentials.j2]: I can think of multiple credentials of the same type... all with the same input values, being all the same credential... it's insane to generate redundant files when post-config will be needed anyway. In the best scenario, the redundant files should be deleted... I see no useful use case with this.

File: roles/filetree_create/templates/current_credentials.j2
Comment (ivarmu) [reply roles/filetree_create/templates/current_credentials.j2]: Same as before.

File: roles/filetree_create/templates/current_schedules.j2
Comment (ivarmu) [reply roles/filetree_create/templates/current_schedules.j2:26]: This is not needed to merge the PR.

**Review by ivarmu — CHANGES_REQUESTED** [review]
The points that doesn't makes sense should be addressed.

**Review by sean-m-sullivan — APPROVED** [review]
This looks great just the few changes and passes

File: roles/filetree_create/templates/current_schedules.j2
Comment (przemkalit) [reply roles/filetree_create/templates/current_schedules.j2:26]: I did not think to customize this part. I won't touch it for now.

File: roles/filetree_create/templates/current_credentials.j2
Comment (przemkalit) [reply roles/filetree_create/templates/current_credentials.j2]: Sometimes it would be good to set global description, because maybe someone wants to use the description as some kind of indicator.

File: roles/filetree_create/templates/current_credential_types.j2
Comment (przemkalit) [reply roles/filetree_create/templates/current_credential_types.j2:6]: Sometimes it would be good to set global description, because maybe someone wants to use the description as some kind of indicator.

File: roles/filetree_create/templates/current_applications.j2
Comment (przemkalit) [reply roles/filetree_create/templates/current_applications.j2:7]: Sometimes it would be good to set global description, because maybe someone wants to use the description as some kind of indicator.

**Comment by przemkalit** [comment]
Ok I've returned from vacation and I've modified the code, I would not change the description customisation but if there is no option to stay as it is, I will do it.

File: roles/filetree_create/templates/current_credentials.j2
Comment (ivarmu) [reply roles/filetree_create/templates/current_credentials.j2]: > Sometimes it would be good to set global description, because maybe someone wants to use the description as some kind of indicator.

Ok, but it would make sense to also allow to set the description from `template_overrides_resources` for completeness.

File: roles/filetree_create/templates/current_credentials.j2
Comment (ivarmu) [inline roles/filetree_create/templates/current_credentials.j2]: There's an extra closing parenthesis at the end of the condition.

**Review by ivarmu — CHANGES_REQUESTED** [review]
I think these will be the last two requested changes ^_^ Good job @przemkalit !!!

Maybe for another future PR, it would be great to have some tests for this staff.

File: roles/filetree_create/templates/current_credentials.j2
Comment (przemkalit) [reply roles/filetree_create/templates/current_credentials.j2]: I don't know why I didn't notice, but thank you very much :).

**Pull request merged**
